src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1418,9 +1418,9 @@ size_t MergeTreeData::clearOldTemporaryDirectories(const MergeTreeDataMergerMuta
             {
                 if (disk->isDirectory(it->path()) && isOldPartDirectory(disk, it->path(), deadline))
                 {
-                    if (merger_mutator.hasTemporaryPart(basename))
+                    if (temporary_parts.contains(basename))
                     {
-                        LOG_WARNING(log, "{} is an active destination for one of merge/mutation (consider increasing temporary_directories_lifetime setting)", full_path);
+                        LOG_WARNING(log, "{} is in use (by merge/mutation/INSERT) (consider increasing temporary_directories_lifetime setting)", full_path);
```

Comment:
Maybe I'm missing something, but how does it protect from concurrent removal during INSERT if no code related to insertions changed in this PR?

Reply:
Egh, it does not.

Can you please also add the following test to your #40031 ?

```sql
CREATE TABLE test_inserts
(
    `key` Int,
    `part` Int                                                                              
)
ENGINE = MergeTree
PARTITION BY part
ORDER BY key
SETTINGS temporary_directories_lifetime = 0, merge_tree_clear_old_temporary_directories_interval_seconds = 0;

INSERT INTO test_inserts                                                                    
SETTINGS max_insert_delayed_streams_for_parallel_write = 100, max_insert_block_size = 1, min_insert_block_size_rows = 1
SELECT
    sleep(1),
    number
FROM numbers(10)
SETTINGS max_insert_delayed_streams_for_parallel_write = 100, max_insert_block_size = 1, min_insert_block_size_rows = 1;

-- Code: 481. DB::Exception: Received from localhost:9000. DB::ErrnoException. DB::ErrnoException: Cannot set modification time for file: ./store/d7c/d7cc238c-e005-4457-8241-4dee47435472/tmp_insert_1_28_28_0, errno: 2, strerror: No such file or directory. (PATH_ACCESS_DENIED) 
```